File: algorithmic_toolbox/week4_divide_and_conquer/2_majority_element/python/majority_element.py
import collections
import sys


def get_majority_element(a: list[int], left: int, right: int) -> int:
    if left == right:
        return -1
    if left + 1 == right:
        return a[left]

    most_common_element = collections.Counter(a).most_common(1)[0]

    return most_common_element[0] if most_common_element[1] > len(a) // 2 else -1


if __name__ == "__main__":
    n, *a = list(map(int, sys.stdin.read().split()))
    if get_majority_element(a, 0, n) != -1:
        print(1)
    else:
        print(0)

File: algorithmic_toolbox/week4_divide_and_conquer/2_majority_element/cpp/majority_element.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using std::vector;

int get_majority_element(vector<int>& a, int left, int right) {
  if (left == right) return -1;
  if (left + 1 == right) return a[left];

  std::unordered_map<int, int> freq;
  for (int i = left; i < right; i++) freq[a[i]]++;

  auto most_common_element = max_element(freq.begin(), freq.end(), [](const auto& a, const auto& b) {
    return a.second < b.second;
  });

  return most_common_element->second > (right - left) / 2 ? most_common_element->first : -1;
}
